### PR TITLE
Fixed chunkSender crash

### PR DESCRIPTION
**Pull request by LogicParrot**
Do not merge yet! The code may change. This is the initial test.
@Seadragon91 does this resolve https://github.com/cuberite/cuberite/issues/2835 and https://github.com/cuberite/cuberite/issues/3345? I cannot reproduce them myself so I couldn't test this.

### CODE REVIEW

File: src/ChunkSender.h [hunk 1]
```diff
@@ -114,6 +114,7 @@ class cChunkSender:
 	cWorld & m_World;
 
 	cCriticalSection  m_CS;
+	cCriticalSection  m_CSClientRemoval;  // This log is held as long as the clients of m_ChunkInfo may be modified
```
Comment (LogicParrot) [inline src/ChunkSender.h]: lock

**Comment by LogicParrot** [comment]
Hypothesis: 
Chunksender grabs a temporary list of client pointers via `std::swap(itr->second.m_Clients, clients);`, now it releases the m_CS lock, enters  `SendChunk()` and begins serializing. Meanwhile, a client is destroyed. Although the client removes itself from the chunkSender, it doesn't remove itself from the temporary list, so `Chunksender` may access a dangling pointer.

File: src/ChunkSender.cpp [hunk 2]
```diff
@@ -204,7 +212,6 @@ void cChunkSender::Execute(void)
 				std::unordered_set<cClientHandle *> clients;
 				std::swap(itr->second.m_Clients, clients);
 				m_ChunkInfo.erase(itr);
-
 				cCSUnlock Unlock(Lock);
```
Comment (LogicParrot) [inline src/ChunkSender.cpp]: This lock has to be freed early, hence the need for the other lock.

**Comment by Seadragon91** [comment]
I am not really sure, have currently 3 crashes under Linux and a deadlock. It looks like the changes from this pull request causes an deadlock.
Look at thread 20.
https://gist.github.com/Seadragon91/f2ef566260670d776086a506dc05ec91

**Comment by tigerw** [comment]
If there is a deadlock, it may be that ChunkSender when attempting to serialise and send, holding both locks, blocking the world thread, enters some protocol method or world method which tries to lock a world mutex.

(*sigh*, in case you've forgotten, may I remind you of our dear Leader, #3115, which uses shared_ptr to ensure correctness)
________________________________

**Comment by Seadragon91** [comment]
@tigerw Please rebase the pull request #3115 on master and I will run my crash script against it :)

**Comment by LogicParrot** [comment]
Last time I checked #3115 was too unstable and easily crashable.

**Comment by LogicParrot** [comment]
@Seadragon91 Are you testing the latest? My first commit can easily deadlock.

**Comment by Seadragon91** [comment]
Joining the server and stopping server, wile player is connected, will result into a hang. Get this under Linux and Windows.

https://gist.github.com/Seadragon91/90d2fef894bce4d32aa647becf38f773

**Comment by Seadragon91** [comment]
From visual studio, 3 threads that have an lock.
```
Cuberite_debug.exe!cCriticalSection::Lock() Zeile 28	 
Cuberite_debug.exe!cCSLock::Lock() Zeile 114	 
Cuberite_debug.exe!cCSLock::cCSLock(cCriticalSection &) Zeile 90	 
Cuberite_debug.exe!cWorld::Stop() Zeile 953	 
Cuberite_debug.exe!cRoot::StopWorlds() Zeile 527	 
Cuberite_debug.exe!cRoot::Start(std::unique_ptr<cSettingsRepositoryInterface,std::default_delete<cSettingsRepositoryInterface> >) Zeile 247	 
Cuberite_debug.exe!UniversalMain(std::unique_ptr<cSettingsRepositoryInterface,std::default_delete<cSettingsRepositoryInterface> >) Zeile 231	 
Cuberite_debug.exe!main(int, char * *) Zeile 548	 
[Externer Code]	 


Cuberite_debug.exe!cCriticalSection::Lock() Zeile 28	 
Cuberite_debug.exe!cCSLock::Lock() Zeile 114	 
Cuberite_debug.exe!cCSUnlock::~cCSUnlock() Zeile 146	 
Cuberite_debug.exe!cChunkSender::Execute() Zeile 217	 
Cuberite_debug.exe!cIsThread::DoExecute() Zeile 75	 
[Externer Code]	 
Cuberite_debug.exe!invoke_thread_procedure(unsigned int(*)(void *), void * const) Zeile 92	 
Cuberite_debug.exe!thread_start<unsigned int (__cdecl*)(void * __ptr64)>(void * const) Zeile 115	 
[Externer Code]	 


Cuberite_debug.exe!cCriticalSection::Lock() Zeile 28	 
Cuberite_debug.exe!cCSLock::Lock() Zeile 114	 
Cuberite_debug.exe!cCSLock::cCSLock(cCriticalSection &) Zeile 90	 
Cuberite_debug.exe!cChunkSender::RemoveClient(cClientHandle *) Zeile 171	 
Cuberite_debug.exe!cWorld::RemoveClientFromChunkSender(cClientHandle *) Zeile 3310	 
Cuberite_debug.exe!cClientHandle::~cClientHandle() Zeile 126	 
[Externer Code]	 
Cuberite_debug.exe!cWorld::TickClients(float) Zeile 1258	 
Cuberite_debug.exe!cWorld::Tick(std::chrono::duration<__int64,std::ratio<1,1000> >, std::chrono::duration<__int64,std::ratio<1,1000> >) Zeile 1053	 
Cuberite_debug.exe!cWorld::cTickThread::Execute() Zeile 112	 
Cuberite_debug.exe!cIsThread::DoExecute() Zeile 75	 
[Externer Code]	 
Cuberite_debug.exe!invoke_thread_procedure(unsigned int(*)(void *), void * const) Zeile 92	 
Cuberite_debug.exe!thread_start<unsigned int (__cdecl*)(void * __ptr64)>(void * const) Zeile 115	 
[Externer Code]
```

**Comment by LogicParrot** [comment]
Alright. I'll look further into this.

`m_evtRemoved` (already in the code) should prevent dangling pointers. Either something is wrong with it, or the problem is not what I guessed.

**Comment by Seadragon91** [comment]
@LogicParrot Just posted something.

**Pull request closed without merging**